libs/contrib/Data/SortedMap.idr

```diff
@@ -319,6 +319,68 @@ export
 mergeLeft : SortedMap k v -> SortedMap k v -> SortedMap k v
 mergeLeft = mergeWith const
 
+treeLeftMost : Tree n k v o -> (k,v)
+treeLeftMost (Leaf x y) = (x,y)
+treeLeftMost (Branch2 x _ _) = treeLeftMost x
+treeLeftMost (Branch3 x _ _ _ _) = treeLeftMost x
+
+treeRightMost : Tree n k v o -> (k,v)
+treeRightMost (Leaf x y) = (x,y)
+treeRightMost (Branch2 _ _ x) = treeRightMost x
+treeRightMost (Branch3 _ _ _ _ x) = treeRightMost x
+
+treeLookupBetween : Ord k => k -> Tree n k v o -> (Maybe (k,v),Maybe (k,v))
+treeLookupBetween k (Leaf k' v) with (k < k')
+  treeLookupBetween k (Leaf k' v) | True = (Nothing, Just (k',v))
+  treeLookupBetween k (Leaf k' v) | False = (Just (k',v), Nothing)
+treeLookupBetween k (Branch2 t1 k' t2) with (k < k')
+  treeLookupBetween k (Branch2 t1 k' t2) | True =
+    let t1r = treeLookupBetween k t1
+    in
+      case t1r of
+        (x,Nothing) => (x,Just $ treeLeftMost t2)
+        _ => t1r
```

Comment:
I'd probably rewrite this the following way (and similarly for the symmetric case):
```suggestion
    let (lower, upper) = treeLookupBetween k t1 in
    (lower, upper <|> pure (treeLeftMost t2))
```